storeliveness: supporter and requester models and transitions

miraradeva opened the pull request:
This commit introduces the Store Liveness algorithm implementation, including the data structures to store the `supportFor` and `supportFrom` state, and the functions to handle Store Liveness messages that update these data structures.

Each of the support data structures can be safely "checked out" for batch update and then "checked in", which allows to make multiple changes, write them to disk (not implemented yet) in batch first, and then reflect the changes in the in-memory data structures. During these batch updates read access to the data structures is not blocked (by holding mutexes or by synchronously waiting for a disk write), enabling quick responses to `SupportFor` and `SupportFrom` calls from the client.

Fixes: #125060

Release note: None

miraradeva committed:
storeliveness: supporter and requester models and transitions

This commit introduces the Store Liveness algorithm implementation,
including the data structures to store the `supportFor` and
`supportFrom` state, and the functions to handle Store Liveness
messages that update these data structures.

Each of the support data structures can be safely updated in batch,
which includes writing the changes to disk (not implemented yet) first,
and then reflecting the changes in the in-memory data structures.
During these batch updates read access to the data structures is not
blocked (by holding mutexes or by synchronously waiting for a disk
write), enabling quick responses to `SupportFor` and `SupportFrom`
calls from the client.

Fixes: #125060

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/127262)
<!-- Reviewable:end -->

nvb reviewed (commented):
> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127262)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @miraradeva)

___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 92 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O21JvJ94SKA_LYXA3ly:-O21JvJ94SKA_LYXA3lz:bdpmpf5) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a9f5b953cca77f261088ec2844c98fdf4903aff/pkg/kv/kvserver/storeliveness/requester_state.go#L92)):*
> ```Go
> func (rs *requesterState) checkOutBatchUpdate() *requesterStateForUpdate {
> 	return &requesterStateForUpdate{
> 		persisted:  rs,
> ```

How does the locking around `*rs` work with this check-out pattern? Should the `RLock` be held on the persisted state for the entire duration of the update to prevent a concurrent update from modifying the persisted state? Or should we only allow a single update at a time?
___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 104 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O21KbI95XkBVC9c7qMZ:-O21KbI95XkBVC9c7qM_:bjuldv7) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a9f5b953cca77f261088ec2844c98fdf4903aff/pkg/kv/kvserver/storeliveness/requester_state.go#L104)):*
> ```Go
> 	rsfu.persisted.mu.Lock()
> 	defer rsfu.persisted.mu.Unlock()
> 	rsfu.persisted.meta = rsfu.inProgress.meta
> ```

Don't we need to update `rsdu.persisted.supportFrom` as well? The naming here is a little awkward, because we aren't actually persisting `supportFrom` state.
___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 130 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O21Ku2iDLmvMkUT40JC:-O21Ku2iDLmvMkUT40JD:bl7uznh) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a9f5b953cca77f261088ec2844c98fdf4903aff/pkg/kv/kvserver/storeliveness/requester_state.go#L130)):*
> ```Go
> 		ss.Expiration.Forward(msg.Expiration)
> 	}
> 	if ss.Epoch < msg.Epoch {
> ```

nit: the three `if` branches in here all look independent. It's a bit of a style thing, but I think it would be helpful for reasoning to make the mutually exclusive cases more clear. For example, making this an `else if` branch (which also ensures that the conditional branches aren't order-dependent). Also optionally moving the MaxEpoch check into this branch.


<!-- Sent from Reviewable.io -->

miraradeva reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127262)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 92 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O21JvJ94SKA_LYXA3ly:-O25LaGG2QQuO8zI7tIk:b-9t5boz) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a9f5b953cca77f261088ec2844c98fdf4903aff/pkg/kv/kvserver/storeliveness/requester_state.go#L92)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

How does the locking around `*rs` work with this check-out pattern? Should the `RLock` be held on the persisted state for the entire duration of the update to prevent a concurrent update from modifying the persisted state? Or should we only allow a single update at a time?
</blockquote></details>

Yeah I was assuming no concurrent updates. I think we do want that assumption but we should probably have some extra protection too.
___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 104 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O21KbI95XkBVC9c7qMZ:-O25IQJI1jQ5vol59qXL:bfmst9i) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a9f5b953cca77f261088ec2844c98fdf4903aff/pkg/kv/kvserver/storeliveness/requester_state.go#L104)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Don't we need to update `rsdu.persisted.supportFrom` as well? The naming here is a little awkward, because we aren't actually persisting `supportFrom` state.
</blockquote></details>

Yes, we should update `supportFrom` as well. The awkward naming actually tripped me up; we don't have to persist `supportFrom` but we should still update the in-memory state. Maybe I'll rename the two states to `checkedIn` and `inProgress` or something like that. 
___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 130 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O21Ku2iDLmvMkUT40JC:-O25JEWr-vrc0INuFmSl:be3ycwj) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a9f5b953cca77f261088ec2844c98fdf4903aff/pkg/kv/kvserver/storeliveness/requester_state.go#L130)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: the three `if` branches in here all look independent. It's a bit of a style thing, but I think it would be helpful for reasoning to make the mutually exclusive cases more clear. For example, making this an `else if` branch (which also ensures that the conditional branches aren't order-dependent). Also optionally moving the MaxEpoch check into this branch.
</blockquote></details>

Agreed. I think doing the following is nice because the second if/elseif is identical to the one we need to do for heartbeat handling (except for the assertions in the comments). We can even extract the common logic away. The heartbeat handling in this draft is actually wrong; the TLA+ spec reminded me that the liveness interval can change and regress the expiration timestamp.
```
if rm.MaxEpoch < msg.Epoch {
    rm.MaxEpoch = msg.Epoch
}
if ss.Epoch == msg.Epoch {
    ss.Expiration.Forward(msg.Expiration)
} else if ss.Epoch < msg.Epoch {
    // assert ss.Epoch == msg.Epoch - 1
    ss.Epoch = msg.Epoch
    // assert msg.Expiration == hlc.Timestamp{}
    ss.Expiration = msg.Expiration
}
```


<!-- Sent from Reviewable.io -->

miraradeva committed:
storeliveness: use maxRequested for heartbeating in TLA+ spec

The TLA+ spec and the store liveness algorithm implementation differ in
the timstamp they use to request support via heartbeats. The algorithm
implementation (in the prototype) first forwards MaxRequested to clock
timestamp + liveness interval, and then uses it as the heartbeat
expiration. The TLA+ spec does the same for updating MaxRequested but
it uses current clock timestamp + liveness interval directly for the
heartbeat expiration. These are usually the same except when
MaxRequested is already greater than clock timestamp + liveness
interval. In this case, in the algorithm implementation, neither
MaxRequested nor the heartbeat expiration regress. In the TLA+ spec,
MaxRequested doesn't regress, but the heartbeat expiration does. This
is not a correctness issue because on the receiver side, the support
provider does not regress the supported timestamp.

This patch brings the TLA+ spec on par with the algorithm
implementation. It also corrects the condition for withdrawing support
to reflect the fact that a support interval's end time is exclusive.

Part of: #125060

Release note: None

nvb reviewed (commented):
> Reviewed 4 of 4 files at r2, 10 of 10 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127262)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @miraradeva)

___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 30 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O3eA0TS5U_d_JTABxLy:-O3eA0TS5U_d_JTABxLz:bdisbsl) ([raw file](https://github.com/cockroachdb/cockroach/blob/540b808f9dfa9eeca35baaa3956e2579ab70501b/pkg/kv/kvserver/storeliveness/requester_state.go#L30)):*
> ```Go
> // - requesterState.addStore(id slpb.StoreIdent)
> // - requesterState.removeStore(id slpb.StoreIdent)
> // NB: getHeartbeatsToSend and handleHeartbeatResponses are not safe for concurrent use.
> ```

Could we clarify that they are not safe for concurrent use with themselves or with each other, but are safe for concurrent use with all other methods?
___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 31 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O3eAGix9nlsyWm-pT58:-O3eAGix9nlsyWm-pT59:byy8fyj) ([raw file](https://github.com/cockroachdb/cockroach/blob/540b808f9dfa9eeca35baaa3956e2579ab70501b/pkg/kv/kvserver/storeliveness/requester_state.go#L31)):*
> ```Go
> // - requesterState.removeStore(id slpb.StoreIdent)
> // NB: getHeartbeatsToSend and handleHeartbeatResponses are not safe for concurrent use.
> ```

nit: this comment isn't currently "attached" to the type def, due to this empty line. That will break godoc.
___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 37 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O3eBTp_4nuGVSPsa4V6:-O3eBTp_4nuGVSPsa4V7:bgl3dw1) ([raw file](https://github.com/cockroachdb/cockroach/blob/540b808f9dfa9eeca35baaa3956e2579ab70501b/pkg/kv/kvserver/storeliveness/requester_state.go#L37)):*
> ```Go
> // respectively, can help ensure safety, but I didn't find an easy way to do
> // this that the linter doesn't complain about.
> // Alternatively, we can add an atomic.Bool updateInProgress to requesterState
> ```

This seems like a nice protection to add.

We could also use a mutex as a form of a refcounter if that would make the role more clear. Something like:
```go
type requesterState struct {
    updateMu syncutil.Mutex
}

func (rs *requesterState) checkOutUpdate() *requesterStateForUpdate {
    if !r.updateMu.TryLock() {
        panic("unsupported concurrent update")
    }
    ...
}

func (rsfu *requesterStateForUpdate) checkInUpdate() {
    defer r.updateMu.Unlock()
    ...
}
```
___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 72 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O3eBoVq-p6BKLm9ZSzg:-O3eBoVq-p6BKLm9ZSzh:b-w4uzwb) ([raw file](https://github.com/cockroachdb/cockroach/blob/540b808f9dfa9eeca35baaa3956e2579ab70501b/pkg/kv/kvserver/storeliveness/requester_state.go#L72)):*
> ```Go
> 		// addStore comes as a result of a SupportFrom call, and there is a
> 		// concurrent update to rs.meta.MaxEpoch (processing heartbeat responses).
> 		// Should rs.mu protect rs.meta as well?
> ```

Having `rs.mu` protect both halves of the data-structure seems reasonable.
___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 101 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O3eCiIP1vXP_DGYZTaE:-O3eCiIP1vXP_DGYZTaF:bci5x9l) ([raw file](https://github.com/cockroachdb/cockroach/blob/540b808f9dfa9eeca35baaa3956e2579ab70501b/pkg/kv/kvserver/storeliveness/requester_state.go#L101)):*
> ```Go
> }
> 
> // getMeta returns the SupporterMeta from the inProgress view; if not present,
> ```

s/SupporterMeta/RequesterMeta/ twice in this sentence.
___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 127 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O3eCFyW8Y3IhOvjASKO:-O3eCFyW8Y3IhOvjASKP:bsn3peu) ([raw file](https://github.com/cockroachdb/cockroach/blob/540b808f9dfa9eeca35baaa3956e2579ab70501b/pkg/kv/kvserver/storeliveness/requester_state.go#L127)):*
> ```Go
> // inProgress view and a checkedIn view that points to the requesterState.
> func (rs *requesterState) checkOutUpdate() *requesterStateForUpdate {
> 	return &requesterStateForUpdate{
> ```

It's may be a premature optimization, but it feels straightfoward to cache a single `requesterStateForUpdate` on `requesterState` so that we don't need to allocate it or its `supportFrom` map each time we want to perform an update. We would then just clear all fields at the end of `checkInUpdate`.

That would also play well with the atomic bool you talked about above.

Same thing with `supporterState.checkOutUpdate`.
___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 138 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O3eDNCr93dqZTRGb3-C:-O3eDNCr93dqZTRGb3-D:b84cs5b) ([raw file](https://github.com/cockroachdb/cockroach/blob/540b808f9dfa9eeca35baaa3956e2579ab70501b/pkg/kv/kvserver/storeliveness/requester_state.go#L138)):*
> ```Go
> // checkInUpdate updates the checkedIn view with any updates from the inProgress
> // view. It persists changes to disk before updating the in-memory objects.
> func (rsfu *requesterStateForUpdate) checkInUpdate() {
> ```

Since we're going to need to persist to disk, we might as well start returning an `error` now so that we immediately build in error handling.

Same thing with `supporterStateForUpdate.checkInUpdate`.
___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 167 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O3eEVxKBV75nDE6_hL2:-O3eEVxKBV75nDE6_hL3:b5zs6y8) ([raw file](https://github.com/cockroachdb/cockroach/blob/540b808f9dfa9eeca35baaa3956e2579ab70501b/pkg/kv/kvserver/storeliveness/requester_state.go#L167)):*
> ```Go
> 		// the prototype they come from a single receive queue). But we can simplify
> 		// the logic here and assume only heartbeat responses are sent to this function.
> 		// This will require separate receive queues for heartbeats and responses.
> ```

I agree that this method which accepts and then ignores `MsgHeartbeat` is a bit awkward. However, I'm not sure separate receive queues are the right answer.

The prototype had a lot of this unfolded with the SupportManager managing a `requesterStateForUpdate` and a `supporterStateForUpdate` directly, allowing the disk writes for all heartbeats and all heartbeat responses could be made durable with a single call to `Commit`. Did you decide that this was not a big enough benefit for the added complexity of having to export the `...ForUpdate` types? Does that still hold once an approach like we have here starts to require the `...ForUpdate.checkInUpdate` methods to manage a `storage.Engine` and `storage.Batch`?
___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 210 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O3eGYME-IOvmhvARqTb:-O3eGYME-IOvmhvARqTc:b-d9itkb) ([raw file](https://github.com/cockroachdb/cockroach/blob/540b808f9dfa9eeca35baaa3956e2579ab70501b/pkg/kv/kvserver/storeliveness/requester_state.go#L210)):*
> ```Go
> 		ss.Expiration.Forward(msg.Expiration)
> 	} else if ss.Epoch < msg.Epoch {
> 		// assert ss.Epoch == msg.Epoch - 1
> ```

Should we add in these assertions? We still don't have a unified assertions framework, which is regrettable. In the meantime, a simple local utility function can get the job done:
```go
// assert panics with the supplied message if the condition does not hold true.
func assert(condition bool, msg string) {
	if !condition {
		panic(msg)
	}
}
```
___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 225 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O3eGzIB4ZdwGnLy8DgZ:-O3eGzIB4ZdwGnLy8Dg_:b-pbd5et) ([raw file](https://github.com/cockroachdb/cockroach/blob/540b808f9dfa9eeca35baaa3956e2579ab70501b/pkg/kv/kvserver/storeliveness/requester_state.go#L225)):*
> ```Go
> ) (heartbeats []slpb.Message) {
> 	// Updating MaxRequested is not a batch update, so it technically doesn't need
> 	// to use the checkIn/checkOut mechanism, but makes the logic simpler.
> ```

"but doing so makes"

Same below.
___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 231 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O3eH4qGBAp9cGhe8zAI:-O3eH4qGBAp9cGhe8zAJ:bcxpeh4) ([raw file](https://github.com/cockroachdb/cockroach/blob/540b808f9dfa9eeca35baaa3956e2579ab70501b/pkg/kv/kvserver/storeliveness/requester_state.go#L231)):*
> ```Go
> 	rs.mu.RLock()
> 	defer rs.mu.RUnlock()
> 	for _, ss := range rs.supportFrom {
> ```

Before iterating, we should presize the slice to avoid resizing as it grows:
```go
heartbeats = make([]slpb.Message, 0, len(rs.supportFrom))
```
___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 259 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O3eHoAQ5dMbiUDuwJ_-:-O3eHoAQ5dMbiUDuwJ_0:bsemnti) ([raw file](https://github.com/cockroachdb/cockroach/blob/540b808f9dfa9eeca35baaa3956e2579ab70501b/pkg/kv/kvserver/storeliveness/requester_state.go#L259)):*
> ```Go
> 	rsfu := rs.checkOutUpdate()
> 	currentEpoch := rsfu.getMeta().MaxEpoch
> 	rsfu.inProgress.meta.MaxEpoch = currentEpoch + 1
> ```

Is this worth a method, if only to fully enumerate all of the ways that the `requestorState` can change?
___
*[`pkg/kv/kvserver/storeliveness/supporter_state.go` line 28 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O3eI7mIE5PkSjzOaUQ6:-O3eI7mIE5PkSjzOaUQ7:bhxea1s) ([raw file](https://github.com/cockroachdb/cockroach/blob/540b808f9dfa9eeca35baaa3956e2579ab70501b/pkg/kv/kvserver/storeliveness/supporter_state.go#L28)):*
> ```Go
> // - supporterState.getSupportFor(id slpb.StoreIdent)
> // NB: handleHeartbeats and withdrawSupport are not safe for concurrent use.
> ```

Same comment about broken godoc.
___
*[`pkg/kv/kvserver/storeliveness/supporter_state.go` line 185 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O3eJDtk5WrxKZib3qeu:-O3eJDtk5WrxKZib3qev:b-jtxxz8) ([raw file](https://github.com/cockroachdb/cockroach/blob/540b808f9dfa9eeca35baaa3956e2579ab70501b/pkg/kv/kvserver/storeliveness/supporter_state.go#L185)):*
> ```Go
> 	func() {
> 		// TODO(mira): do we need to hold the RLock() while iterating over
> 		// ss.supportFor? I think so.
> ```

We aren't expecting anyone to mutate `supporterState` while we have an update checked out, so we shouldn't need to. Same thing with `getHeartbeatsToSend`.

If we did, we'd need to `RLock` in places like `supporterStateForUpdate.getSupportFor` and `requesterStateForUpdate.getSupportFrom`. We actually are doing that, but I don't think we want to. We aren't in `supporterStateForUpdate.getMeta` or `requesterStateForUpdate.getMeta`.
___
*[`pkg/kv/kvserver/storeliveness/supporter_state.go` line 215 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O3eLf4S1O-62EWkFS7z:-O3eLf4S1O-62EWkFS8-:b-vb9ne6) ([raw file](https://github.com/cockroachdb/cockroach/blob/540b808f9dfa9eeca35baaa3956e2579ab70501b/pkg/kv/kvserver/storeliveness/supporter_state.go#L215)):*
> ss.Expiration.Less(now.ToTimestamp())

We had this logic in the prototype as well, but I'm not sure it's quite right. We usually treat expirations as exclusive, like in `roachpb.Lease`. So if `ss.Expiration == now.ToTimestamp()`, I believe we should be able to withdraw support. Do you mind double checking that and clarifying in the field's comment?
___
*[`pkg/kv/kvserver/storeliveness/store_liveness_test.go` line 155 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O3eOpP_3a0DuAzW8Qwl:-O3eOpP_3a0DuAzW8Qwm:bje4vyc) ([raw file](https://github.com/cockroachdb/cockroach/blob/540b808f9dfa9eeca35baaa3956e2579ab70501b/pkg/kv/kvserver/storeliveness/store_liveness_test.go#L155)):*
> ```Go
> 	var wallTimeSecs int64
> 	d.ScanArgs(t, "now", &wallTimeSecs)
> 	wallTime := wallTimeSecs * 1000000000
> ```

nit: `s/1000000000/int64(time.Second)/` here and below.
___
*[`pkg/kv/kvserver/storeliveness/store_liveness_test.go` line 172 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O3eQ_ECEahxhgEi8QB3:-O3eQ_ECEahxhgEi8QB4:bdmrbmy) ([raw file](https://github.com/cockroachdb/cockroach/blob/540b808f9dfa9eeca35baaa3956e2579ab70501b/pkg/kv/kvserver/storeliveness/store_liveness_test.go#L172)):*
> ```Go
> 		}
> 		var msgType int
> 		d.ScanArgs(t, "type", &msgType)
> ```

Accepting a string here ("MsgHeartbeat" or "MsgHeartbeatResp") instead of an integer would make the test files more readable. Something like:
```go
var msgType slpb.MessageType
var msgTypeStr string
d.ScanArgs(t, "type", &msgTypeStr)
switch {
case slpb.MsgHeartbeat.String():
    msg = slpb.MsgHeartbeat
case slpb.MsgHeartbeatResp.String():
    msg = slpb.MsgHeartbeatResp
default:
    d.Fatalf(t, "unexpected \"type\", found %s", msgTypeStr)
}
```


<!-- Sent from Reviewable.io -->

miraradeva reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127262)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 127 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O3eCFyW8Y3IhOvjASKO:-O3hpX0QEIblsFMQzrgX:b-rnldno) ([raw file](https://github.com/cockroachdb/cockroach/blob/540b808f9dfa9eeca35baaa3956e2579ab70501b/pkg/kv/kvserver/storeliveness/requester_state.go#L127)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

It's may be a premature optimization, but it feels straightfoward to cache a single `requesterStateForUpdate` on `requesterState` so that we don't need to allocate it or its `supportFrom` map each time we want to perform an update. We would then just clear all fields at the end of `checkInUpdate`.

That would also play well with the atomic bool you talked about above.

Same thing with `supporterState.checkOutUpdate`.
</blockquote></details>

Yeah I like this idea, and I incorporated it. But I wonder: if we're moving the `requesterStateForUpdate` to be stored on the `requesterState`, then does it still make sense for the `requesterStateForUpdate` to have a reference to `requesterState` (that's the `checkedIn` pointer)? Maybe we can have something like this:

```
type supporterState struct {
    meta slpb.SupporterMeta
    supportFor map[slpb.StoreIdent]slpb.SupportState
}

type supporterStateForUpdate { // or better name
    // mu protects checkedIn
    mu syncutil.RWMutex 
    checked in *supporterState

    // updateMu ensures there is only one update in progress.
    updateMu syncutil.Mutex
    inProgress *supporterState
}
```

The `supporterStateForUpdate` will be stored in `SupportManager`. It might be slightly more error-prone mutating the correct `supporterState` (`checkedIn` or `inProgress`).
___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 167 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O3eEVxKBV75nDE6_hL2:-O3gy9_6BVhfbG75xBu-:b-6xgknu) ([raw file](https://github.com/cockroachdb/cockroach/blob/540b808f9dfa9eeca35baaa3956e2579ab70501b/pkg/kv/kvserver/storeliveness/requester_state.go#L167)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I agree that this method which accepts and then ignores `MsgHeartbeat` is a bit awkward. However, I'm not sure separate receive queues are the right answer.

The prototype had a lot of this unfolded with the SupportManager managing a `requesterStateForUpdate` and a `supporterStateForUpdate` directly, allowing the disk writes for all heartbeats and all heartbeat responses could be made durable with a single call to `Commit`. Did you decide that this was not a big enough benefit for the added complexity of having to export the `...ForUpdate` types? Does that still hold once an approach like we have here starts to require the `...ForUpdate.checkInUpdate` methods to manage a `storage.Engine` and `storage.Batch`?
</blockquote></details>

Yeah that's a good point. My goal was to keep as much of the algorithm logic here, in the requester and supporter states, and have the `SupportManager` only deal with orchestrating things like the receive queue, the main loop that drives the algorithm logic. 

The single `Commit` is definitely nice and I went back and forth about it a few times. The current approach will need two `Commit`s, one for the requester state update and one for the receive state update, which I thought was a reasonable tradeoff for a better separation of concerns. But managing `storage.Engine` and `storage.Batch` here also doesn't sound great. 

Keeping all the pieces of the algorithm needed for correctness here is not possible anyway, because a lot of the clock-related invariants will only make sense in the context of the `SupportManager`, where the clock reference lives. I think I'd keep these here for now, if only to be able to test the correctness before the `SupportManager` is ready. But will add a TODO to consider moving them to the `SupportManager`. 
___
*[`pkg/kv/kvserver/storeliveness/supporter_state.go` line 185 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O3eJDtk5WrxKZib3qeu:-O3h3s7Y5rhlPPo0Na_n:b2cea7r) ([raw file](https://github.com/cockroachdb/cockroach/blob/540b808f9dfa9eeca35baaa3956e2579ab70501b/pkg/kv/kvserver/storeliveness/supporter_state.go#L185)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We aren't expecting anyone to mutate `supporterState` while we have an update checked out, so we shouldn't need to. Same thing with `getHeartbeatsToSend`.

If we did, we'd need to `RLock` in places like `supporterStateForUpdate.getSupportFor` and `requesterStateForUpdate.getSupportFrom`. We actually are doing that, but I don't think we want to. We aren't in `supporterStateForUpdate.getMeta` or `requesterStateForUpdate.getMeta`.
</blockquote></details>

You're right, we don't expect `supporterState.supportFor` to be mutated while an update is checked out. But I think `getHeartbeatsToSend` (and generally anything reading from `requesterState.supportFrom`) is different because a store can be added to `requesterState.supportFrom` concurrently with an ongoing update; e.g. as part of `SupportFrom` a new store.
___
*[`pkg/kv/kvserver/storeliveness/supporter_state.go` line 215 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O3eLf4S1O-62EWkFS7z:-O3hE0Lw51-HZvK5mdlX:b-6ol75a) ([raw file](https://github.com/cockroachdb/cockroach/blob/540b808f9dfa9eeca35baaa3956e2579ab70501b/pkg/kv/kvserver/storeliveness/supporter_state.go#L215)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> ss.Expiration.Less(now.ToTimestamp())

We had this logic in the prototype as well, but I'm not sure it's quite right. We usually treat expirations as exclusive, like in `roachpb.Lease`. So if `ss.Expiration == now.ToTimestamp()`, I believe we should be able to withdraw support. Do you mind double checking that and clarifying in the field's comment?
</blockquote></details>

Yup, `roachpb.Lease` states this about the expiration field: `This is an exclusive value, i.e. the lease is valid in [start, expiration)`. We can do the same here. I already have a test for it.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
Haven't given the latest iteration a full review yet, but flushing comments before some meetings.

> Reviewed 11 of 11 files at r4, 1 of 10 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-:-O3nLyQMBRN4I1NL9kpQ:bldnjjo)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @miraradeva)

___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 27 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O3nIYeW5CQp5ATH9cBz:-O3nIYeW5CQp5ATH9cC-:b-hep5kb) ([raw file](https://github.com/cockroachdb/cockroach/blob/90ed323320f19c353100467c26a81b7eb7b098f2/pkg/kv/kvserver/storeliveness/requester_state.go#L27)):*
> ```Go
> //   - removeStore(id slpb.StoreIdent)
> //   - checkOutUpdate()
> //     getHeartbeatsToSend(now hlc.Timestamp,	interval time.Duration)
> ```

nit: stray table before "interval".
___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 139 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O3nKsH72VkzOHvcCsK-:-O3nKsH72VkzOHvcCsK0:bcwvdvw) ([raw file](https://github.com/cockroachdb/cockroach/blob/90ed323320f19c353100467c26a81b7eb7b098f2/pkg/kv/kvserver/storeliveness/requester_state.go#L139)):*
> ```Go
> // inProgress view and a checkedIn view that points to the requesterState
> // (if one doesn't exist already).
> func (rsh *requesterStateHandler) checkOutUpdate() {
> ```

The latest diff removed the `*requesterStateForUpdate` return value and moved the mutation methods onto `requesterStateHandler`. Why do that? It seems more error prone, because now it's possible to call mutating methods like `handleHeartbeatResponse` without having checked out the update struct. Shouldn't these methods be on the update struct and require a `checkOutUpdate` call?
___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 140 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O3nKHezCsdBc96sope1:-O3nKHezCsdBc96sope2:b-y2cfqh) ([raw file](https://github.com/cockroachdb/cockroach/blob/90ed323320f19c353100467c26a81b7eb7b098f2/pkg/kv/kvserver/storeliveness/requester_state.go#L140)):*
> ```Go
> // (if one doesn't exist already).
> func (rsh *requesterStateHandler) checkOutUpdate() {
> 	if !rsh.updateMu.TryLock() {
> ```

Did you end up giving the `atomic.Pointer[requesterStateForUpdate]` suggestion a try? Looking at this again, I do think it will lead to clearer logic. We'd `Swap(nil)` to check out and `Swap(val)` to check it back in.
___
*[`pkg/kv/kvserver/storeliveness/store_liveness_test.go` line 38 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O3nK6-w-A-MwSeYxMzO:-O3nK6-w-A-MwSeYxMzP:bcfu17y) ([raw file](https://github.com/cockroachdb/cockroach/blob/90ed323320f19c353100467c26a81b7eb7b098f2/pkg/kv/kvserver/storeliveness/store_liveness_test.go#L38)):*
> ```Go
> 	datadriven.Walk(
> 		t, datapathutils.TestDataPath(t), func(t *testing.T, path string) {
> 			ss := supporterStateHandler{
> ```

Should we add a constructor for `supporterStateHandler` and `requesterStateHandler` so that we don't need to reach inside and make sure to init all of the fields in multiple places?

This would also play well with the `atomic.Pointer` suggestion above.


<!-- Sent from Reviewable.io -->

miraradeva reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127262)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 139 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O3nKsH72VkzOHvcCsK-:-O3nNLQt63dshdykfv_A:b-wt5s4x) ([raw file](https://github.com/cockroachdb/cockroach/blob/90ed323320f19c353100467c26a81b7eb7b098f2/pkg/kv/kvserver/storeliveness/requester_state.go#L139)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

The latest diff removed the `*requesterStateForUpdate` return value and moved the mutation methods onto `requesterStateHandler`. Why do that? It seems more error prone, because now it's possible to call mutating methods like `handleHeartbeatResponse` without having checked out the update struct. Shouldn't these methods be on the update struct and require a `checkOutUpdate` call?
</blockquote></details>

I was hoping to keep `requesterStateForUpdate` hidden from the `SupportManager` and used only internally in `requesterStateHandler`. I did add a check for holding `updateMu` when calling mutating methods. Either way seems good to me. 
___
*[`pkg/kv/kvserver/storeliveness/requester_state.go` line 140 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O3nKHezCsdBc96sope1:-O3nNCmp-rO2o0Pp8BGQ:b-b6x09b) ([raw file](https://github.com/cockroachdb/cockroach/blob/90ed323320f19c353100467c26a81b7eb7b098f2/pkg/kv/kvserver/storeliveness/requester_state.go#L140)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Did you end up giving the `atomic.Pointer[requesterStateForUpdate]` suggestion a try? Looking at this again, I do think it will lead to clearer logic. We'd `Swap(nil)` to check out and `Swap(val)` to check it back in.
</blockquote></details>

Yeah I can try it. 
___
*[`pkg/kv/kvserver/storeliveness/store_liveness_test.go` line 38 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-O3nK6-w-A-MwSeYxMzO:-O3nNJPj4x2GSahgI48E:bmiskcq) ([raw file](https://github.com/cockroachdb/cockroach/blob/90ed323320f19c353100467c26a81b7eb7b098f2/pkg/kv/kvserver/storeliveness/store_liveness_test.go#L38)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we add a constructor for `supporterStateHandler` and `requesterStateHandler` so that we don't need to reach inside and make sure to init all of the fields in multiple places?

This would also play well with the `atomic.Pointer` suggestion above.
</blockquote></details>

Yes, good idea.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 6 of 9 files at r6, 3 of 3 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127262#-:-O4BdBgo0wNGHxlYduAR:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani)



<!-- Sent from Reviewable.io -->

miraradeva commented:
TFTR!

bors r=nvanbenschoten

nvanbenschoten commented:
Looks like bors crashed.

bors r+
